cmd/crowdsec/output.go

```diff
@@ -1,31 +1,103 @@
 package main
 
 import (
+	"context"
 	"fmt"
+	"net/url"
+	"sync"
+	"time"
 
+	"github.com/go-openapi/strfmt"
+	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 
+	"github.com/crowdsecurity/crowdsec/pkg/apiclient"
+	"github.com/crowdsecurity/crowdsec/pkg/csconfig"
 	leaky "github.com/crowdsecurity/crowdsec/pkg/leakybucket"
+	"github.com/crowdsecurity/crowdsec/pkg/models"
 	"github.com/crowdsecurity/crowdsec/pkg/parser"
 	"github.com/crowdsecurity/crowdsec/pkg/types"
-	"github.com/davecgh/go-spew/spew"
 )
 
-func runOutput(input chan types.Event, overflow chan types.Event, holders []leaky.BucketFactory, buckets *leaky.Buckets,
-	poctx parser.UnixParserCtx, ponodes []parser.Node) error {
+func dedupAlerts(alerts []types.RuntimeAlert) ([]*models.Alert, error) {
+
+	var dedupCache []*models.Alert
+
+	for idx, alert := range alerts {
+		log.Debugf("alert %d/%d", idx, len(alerts))
+		/*if we have more than one source, we need to dedup */
+		if len(alert.Sources) == 0 || len(alert.Sources) == 1 {
+			dedupCache = append(dedupCache, alert.Alert)
+			continue
+		}
+		for k, src := range alert.Sources {
+			refsrc := *alert.Alert //copy
+			log.Debugf("source[%s]", k)
+			refsrc.Source = &src
+			dedupCache = append(dedupCache, &refsrc)
+		}
+	}
+	if len(dedupCache) != len(alerts) {
+		log.Infof("went from %d to %d alerts", len(alerts), len(dedupCache))
+	}
+	return dedupCache, nil
+}
+
+func PushAlerts(alerts []types.RuntimeAlert, client *apiclient.ApiClient) error {
+	ctx := context.Background()
+	alertsToPush, err := dedupAlerts(alerts)
+
+	if err != nil {
+		return errors.Wrap(err, "failed to transform alerts for api")
+	}
+	resp, _, err := client.Alerts.Add(ctx, alertsToPush)
+	if err != nil {
+		return errors.Wrap(err, "failed sending alert to apil")
+	}
+	return nil
+}
+
+func runOutput(input chan types.Event, overflow chan types.Event, buckets *leaky.Buckets,
+	postOverflowCTX parser.UnixParserCtx, postOverflowNodes []parser.Node, apiConfig csconfig.ApiCredentialsConfig) error {
+
+	var err error
+	ticker := time.NewTicker(1 * time.Second)
+
+	var cache []types.RuntimeAlert
+	var cacheMutex sync.Mutex
+
+	apiclient.BaseURL, err = url.Parse(apiConfig.Url)
+	if err != nil {
+		return fmt.Errorf("unable to parse api url '%s': %s", apiConfig.Url, err)
+	}
+	password := strfmt.Password(apiConfig.Password)
+	t := &apiclient.JWTTransport{
+		MachineID: &apiConfig.Login,
+		Password:  &password,
+		Scenarios: []string{"aaaaaa", "bbbbb"},
+	}
+	Client := apiclient.NewClient(t.Client())
 
 LOOP:
 	for {
 		select {
+		case <-ticker.C:
+			if len(cache) > 0 {
+				cacheMutex.Lock()
+				cachecopy := cache
+				newcache := make([]types.RuntimeAlert, 0)
+				cache = newcache
+				if err := PushAlerts(cachecopy, Client); err != nil {
+					log.Errorf("while pushing to api : %s", err)
+				}
+				cacheMutex.Unlock()
```

Comment:
we can move the unlock before the pushAlerts to avoid locking the routine during the long operation